proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+        revision: String,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        _ range: Range<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        _ range: ClosedRange<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        url: String,
+        exact version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    // MARK: Registry 
+
+    public static func package(
+        id: String,
+        from version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        exact version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        _ range: Range<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        _ range: ClosedRange<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+}
+```
+
+### Default traits
+
+Default traits allow package authors to define a set of traits that they think
+cater to the majority use-cases of the package. When choosing the initial
+default traits or adding a new default trait it is important to consider that
+removing a default trait is a SemVer major change since it can potentially
+remove APIs.
+
+### Trait specific command line options for `swift build/run`
+
+When executing one of `swift build/run` options can be passed to control
+which traits for the root package are enabled:
+
+- `--traits` _TRAITS_: Enables the passed traits of the package. Multiple traits can be
+  specified by providing a comma separated list e.g. `--traits Trait1,Trait2`.
+- `--enable-all-traits`: Enables all traits of the package.
+- `--disable-default-traits`: Disables all default traits of the package.
+
+
+### Trait limitations
+
+To prevent abuse, limit the complexity and make sure it integrates with the
+compiler a few limitations are imposed.
+
+#### Number of traits
+
+[Other
+ecosystems](https://blog.rust-lang.org/2023/10/26/broken-badges-and-23k-keywords.html)
+have shown that a large number of traits can have significant impact on
+registries and dependency managers. To avoid such a scenario an initial maximum
+number of traits of 300 per package is imposed. This can be revisited later once
+traits have been used in the ecosystem extensively.
+
+### Allowed characters for trait names
+
+Since traits can show up both in the `Package.swift` and in source code when
+checking if a trait is enabled, the allowed characters for a trait name have to
+be restricted. The following rules are enforced on trait names:
+
+- The first character must be a [Unicode XID start
+  character](https://unicode.org/reports/tr31/#Figure_Code_Point_Categories_for_Identifier_Parsing)
+  (most letters), a digit, or `_`.
```

Comment:
Why do we allow a digit as the first character? This seems counter to other naming conventions we have.

Reply:
AFAIK this follows the exact same enforcements as `targets` or `products` in the `Package.swift`. I agree that a good convention is to not use digits. The Unicode XID standard is specifically defined for parsers and something that we enforce in our tooling.

Reply:
I don't think that's a good approach. Why not restrict character names to what's possible as Swift identifiers? That would allow people to define traits as enum cases, which would improve diagnostics significantly. Currently if someone makes type in trait's name, we have to implement special diagnostic handling (which is hard to implement well for package manifests), and you'll get that diagnostic only when you try to resolve instead of getting a usual error inline in an IDE about nonexistent enum case.

Reply:
I think the `Unicode XID start` and `Unicode XID continue` character sets are either matching what `Swift identifier`s are allowed to be or a subset. However, I think this is worth confirming with somebody from the compiler.

Reply:
The [official grammar for identifiers](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/summaryofthegrammar#:~:text=Grammar%20of%20an%20identifier) supports this, if I read it correctly.

Reply:
> Why not restrict character names to what's possible as Swift identifiers?

Agreed with this, Swift identifier grammar seems like a good template.

> That would allow people to define traits as enum cases

On that note - I wonder if there could be initializers for traits that take a generic `RawRepresentable` value instead of the string?

That'd allow:

```swift
enum MyTraits: String {
    case foo
    case bar
}

let trait: Trait = .trait(name: MyTraits.foo)

// Without this, one has to do:
let trait: Trait = .trait(name: MyTraits.foo.rawValue)
```

Reply:
This is expanded in the alternatives considered sections. I explored using enums but the really nice spelling isn't possible right now since `Package` can't become generic of a `Trait: RawRepresentable`. Furthermore, nothing in the manifest is typed right now. Everything is `String` based. That's why this pitch uses `String`s. A future evolution can check how we can make everything strongly typed but it feels wrong if just traits are typed but nothing else is.

Reply:
I think I'm only asking for an initializer, not making the container types generic. So something like this:

```swift
struct Trait {
    var name: String
    // ...
    static func trait(name: some RawRepresentable<String>) -> Self {
        .init(name: name.rawValue)
    }
}
```

> nothing in the manifest is typed right now

That's fair, for sure. But traits are very enum-y conceptually, so I'd especially prefer using enums here, even more than for target names, etc.

Reply:
Target, products and dependencies are also enum like. I don't think we should add the `RawRepresentable` overload since none of the other types have it. It also doesn't add much more value since as you showed you can use the `.rawValue` easily.